File: MacOS/Type4Me/ASR/Providers/VolcanoASRConfig.swift
import Foundation

struct VolcanoASRConfig: ASRProviderConfig, Sendable {

    static let provider = ASRProvider.volcano
    static var displayName: String { L("火山引擎 (Doubao)", "Volcano (Doubao)") }

    static var credentialFields: [CredentialField] {[
        CredentialField(key: "appKey", label: "App Key", placeholder: "APPID", isSecure: false, isOptional: false, defaultValue: ""),
        CredentialField(key: "accessKey", label: "Access Token", placeholder: L("访问令牌", "Access token"), isSecure: true, isOptional: false, defaultValue: ""),
    ]}

    let appKey: String
    let accessKey: String
    let resourceId: String
    let uid: String

    init?(credentials: [String: String]) {
        guard let appKey = credentials["appKey"], !appKey.isEmpty,
              let accessKey = credentials["accessKey"], !accessKey.isEmpty
        else { return nil }
        self.appKey = appKey
        self.accessKey = accessKey
        self.resourceId = credentials["resourceId"]?.isEmpty == false
            ? credentials["resourceId"]!
            : "volc.bigasr.sauc.duration"
        self.uid = ASRIdentityStore.loadOrCreateUID()
    }

    func toCredentials() -> [String: String] {
        ["appKey": appKey, "accessKey": accessKey, "resourceId": resourceId]
    }

    var isValid: Bool {
        !appKey.isEmpty && !accessKey.isEmpty
    }
}

File: Windows/Type4Me/ASR/Providers/VolcanoASRConfig.cs
using Type4Me.Localization;

namespace Type4Me.ASR.Providers;

public sealed class VolcanoASRConfig
{
    public const string DefaultResourceId = "volc.bigasr.sauc.duration";
    public const string FlashResourceId = "volc.bigasr.auc_turbo";

    public string ApiKey { get; }
    public string ResourceId { get; }
    public string Uid { get; }

    public static ASRProvider Provider => ASRProvider.Volcano;
    public static string DisplayName => Loc.L("火山引擎 (Doubao)", "Volcano (Doubao)");

    public static CredentialField[] CredentialFields =>
    [
        new() { Key = "apiKey", Label = "API Key", Placeholder = Loc.L("新版控制台 API Key", "API Key from new console"), IsSecure = true, IsOptional = false },
        new() { Key = "resourceId", Label = Loc.L("资源 ID", "Resource ID"), Placeholder = DefaultResourceId, IsSecure = false, IsOptional = true },
    ];

    private VolcanoASRConfig(string apiKey, string resourceId, string uid)
    {
        ApiKey = apiKey;
        ResourceId = resourceId;
        Uid = uid;
    }

    public static VolcanoASRConfig? TryCreate(Dictionary<string, string> credentials)
    {
        // New-console single API key. Accept legacy "accessKey" as a fallback for migration.
        if (!credentials.TryGetValue("apiKey", out var apiKey) || string.IsNullOrWhiteSpace(apiKey))
        {
            credentials.TryGetValue("accessKey", out apiKey);
        }
        if (string.IsNullOrWhiteSpace(apiKey))
            return null;

        var resourceId = credentials.GetValueOrDefault("resourceId");
        if (string.IsNullOrWhiteSpace(resourceId))
            resourceId = DefaultResourceId;

        var uid = Services.ASRCustomizationStorage.LoadOrCreateUID();

        return new VolcanoASRConfig(apiKey.Trim(), resourceId.Trim(), uid);
    }

    public Dictionary<string, string> ToCredentials() => new()
    {
        ["apiKey"] = ApiKey,
        ["resourceId"] = ResourceId,
    };

    public bool IsValid => !string.IsNullOrEmpty(ApiKey);
}
